src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/nodes/plugins/ffi/PrimExternalCallNode.java

```diff
@@ -0,0 +1,79 @@
+package de.hpi.swa.trufflesqueak.nodes.plugins.ffi;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import de.hpi.swa.trufflesqueak.exceptions.PrimitiveFailed;
+import de.hpi.swa.trufflesqueak.nodes.primitives.AbstractPrimitiveNode;
+import de.hpi.swa.trufflesqueak.util.FrameAccess;
+import de.hpi.swa.trufflesqueak.util.NFIUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class PrimExternalCallNode extends AbstractPrimitiveNode {
+    final String moduleName;
+    final String functionName;
+    final int numReceiverAndArguments;
+    static Map<String, Object> loadedLibraries = new HashMap<>();
+
+    public PrimExternalCallNode(String moduleName, String functionName, int numReceiverAndArguments) {
+        this.moduleName = moduleName;
+        this.functionName = functionName;
+        this.numReceiverAndArguments = numReceiverAndArguments;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
```

Comment:
There's too much code in this method that is not designed for PE, so we need to put some of it behind `@TruffleBoundary`, maybe even materialize the frame all the time. This probably causes the native builds to fail.